packages/gatsby-plugin-benchmark-reporting/gatsby-node.js

```diff
@@ -0,0 +1,198 @@
+      benchmarkStart: 0, // Start of benchmark itself
+      preInit: 0, // Gatsby onPreInit life cycle
+      preBootstrap: 0, // Gatsby onPreBootstrap life cycle
+      preBuild: 0, // Gatsby onPreBuild life cycle
+      postBuild: 0, // Gatsby onPostBuild life cycle
+      benchmarkEnd: 0, // End of benchmark itself
+    }
+    this.started = false
+  }
+
+  getData() {
+    for (const key in this.timestamps) {
+      this.timestamps[key] = Math.floor(this.timestamps[key])
+    }
+
+    const pageCount = parseInt(execSync("find . | grep page-data | wc -l", {
+      cwd: path.resolve("public/page-data"),
+      encoding: "utf8",
+    }).trim(), 10)
+
+    return {
+      time: this.localTime,
+      sessionId: uuidv4(),
+      pageCount,
+      timestamps: this.timestamps,
+    }
+  }
+
+  markStart() {
+    if (this.started) {
+      reportError(
+        "gatsby-plugin-benchmark-reporting: ",
+        new Error("Error: Should not call markStart() more than once")
+      )
+      process.exit(1)
+    }
+    this.timestamps.benchmarkStart = performance.now()
+    this.started = true
+  }
+
+  markDataPoint(name) {
+    if (BENCHMARK_REPORTING_URL) {
+      if (!(name in this.timestamps)) {
+        reportError(
+          `Attempted to record a timestamp with a name (\`${name}\`) that wasn't expected`
+        )
+        process.exit(1)
+      }
+    }
+    this.timestamps[name] = performance.now()
+  }
+
+  async markEnd() {
+    if (!this.timestamps.benchmarkStart) {
+      reportError(
+        "gatsby-plugin-benchmark-reporting:",
+        new Error("Error: Should not call markEnd() before calling markStart()")
+      )
+      process.exit(1)
+    }
+    this.timestamps.benchmarkEnd = performance.now()
+    return this.flush()
+  }
+
+  async flush() {
+    const data = this.getData()
+    const json = JSON.stringify(data)
+
+    reportInfo("Submitting data: " + json)
+
+    if (BENCHMARK_REPORTING_URL) {
+      reportInfo("Flushing benchmark data to remote server...")
+
+      let lastStatus = 0
+      this.flushing = nodeFetch(`${BENCHMARK_REPORTING_URL}`, {
+        method: `POST`,
+        headers: {
+          "content-type": `application/json`,
+          // "user-agent": this.getUserAgent(),
+        },
+        body: json,
+      }).then(res => {
+        lastStatus = res.status
+        if (lastStatus === 500) {
+          reportError(
+            "Response error",
+            new Error("Server responded with a 500 error")
+          )
+          process.exit(1)
+        }
+        this.flushed = true
+        // Note: res.text returns a promise
+        return res.text()
+      })
+
+      this.flushing.then(text =>
+        reportInfo(`Server response: ${lastStatus}: ${text}`)
+      )
+
+      return this.flushing
+    }
+
+    // ENV var had no reporting end point
+
+    this.flushing = Promise.resolve()
+    this.flushed = true
+  }
+}
+
+process.on(`exit`, async () => {
```

Comment:
https://nodejs.org/api/process.html#process_event_exit :

```
Listener functions must only perform synchronous operations. The Node.js process will exit immediately after calling the 'exit' event listeners causing any additional work still queued in the event loop to be abandoned. 
```

There is `beforeExit` event that allow `async` stuff, but not exactly sure about details here - https://nodejs.org/api/process.html#process_event_beforeexit

Reply:
Ok, fair enough :)

I'll just log and process.exit immediately

Reply:
Up to you, if you decide if this is a blocker (I don't have enough context), just wanted to point potential issues with this ;)

Reply:
Fixed :)